prow/sidecar/run.go

```diff
@@ -208,17 +211,31 @@ func (o Options) preUpload() {
 	}
 }
 
-func (o Options) doUpload(ctx context.Context, spec *downwardapi.JobSpec, passed, aborted bool, metadata map[string]interface{}, logReaders map[string]io.Reader) error {
+func (o Options) doUpload(ctx context.Context, spec *downwardapi.JobSpec, passed, aborted bool, metadata map[string]interface{}, logReaders map[string]io.Reader, logFile *os.File) error {
 	startTime := time.Now()
 	logrus.Info("Starting to upload")
-	defer func() { logrus.WithField("duration", time.Since(startTime).String()).Info("Finished uploading") }()
-
 	uploadTargets := make(map[string]gcs.UploadFunc)
 
+	defer func() {
+		logrus.WithField("duration", time.Since(startTime).String()).Info("Finished uploading")
+	}()
+
 	for logName, reader := range logReaders {
 		uploadTargets[logName] = gcs.DataUpload(reader)
 	}
 
+	logFile.Sync()
+	logrus.SetOutput(os.Stderr)
+	logFileName := logFile.Name()
+	logFile.Close()
+
+	f, err := os.Open(logFileName)
+	if err != nil {
+		logrus.Fatalf("Could not open log file: %v", err)
+	}
+	uploadTargets[logFileName] = gcs.DataUpload(bufio.NewReader(f))
+	f.Close()
```

Comment:
We can't close the file here. On the previous line we didn't actually read the file, we just prepared a reader that will be consumed by `o.GcsOptions.Run(...)`. That consumer function is responsible for closing the file once the data has been read.